### PR TITLE
Validate PMIs, ignore duplicates

### LINKED ISSUE #309: Behavior for multiple modifiers for the same payment method is not defined
Currently the specification defines a `PaymentDetailsModifier` sequence which can be used to apply modifications to the total and display items as appropriate for the currently selected payment method:

https://w3c.github.io/browser-payment-api/#dom-paymentdetailsmodifier

However, the specification does not define the user agent behavior in the case where this sequence contains multiple `PaymentDetailsModifier` objects for the same payment method. This needs to be defined, otherwise user agents risk implementing differing behaviors that could result in incorrect amounts being charged.

I would suggest since this is a sequence stating the first entry to match a payment method will be used as its modifier.

Alternatively, we could say supplying multiple modifiers for a single payment instrument is not allowed and throw a `TypeError`.

### LINKED ISSUE #464: validate supportedMethods
The spec currently specifies: 

```
dictionary PaymentMethodData {
    DOMString supportedMethods;
};
```

But those are not actually strings: they are either enums (e.g., "basic-card") or USVStrings for URLs. This causes problems when doing URL comparisons as required by the Payment Method Identifiers spec (which calls into the URL spec).

### PR DESCRIPTION
**Pull request by marcoscaceres**
* Validate supportedMethods (closes #464)
* Defines behavior for duplicate PMIs in constructor and updateWith() (closes #309)

* [x]  ctor PMI check tests: https://github.com/w3c/web-platform-tests/pull/6816
* [x]  updateWith() PMI validity tests: https://github.com/w3c/web-platform-tests/pull/7021

***
[Preview](https://s3.amazonaws.com/pr-preview/w3c/payment-request/validate_and_ignore.html) | [Diff](https://s3.amazonaws.com/pr-preview/w3c/payment-request/76bd629...efde8c3.html)

### CODE REVIEW

**Comment by marcoscaceres** [comment]
Working on tests... will update OP as I send them to the WPT repo.

**Comment by adrianhopebailie** [comment]
@zkoch and @adrianba can you confirm that this matches the implementation on Edge and Chrome.

I read the proposed text to not allow this:

```javascript
new PaymentRequest(
  [
    {
      supportedMethods: "basic-card",
      data: {
        supportedTypes: ["debit"],
        supportedNetworks: ["visa", "amex"],
      }
    },
    {
      supportedMethods: "basic-card",
      data: {
         supportedTypes: ["credit"],
      }
    },
  ],
  details
);
```
...implying, I accept credit cards from all networks but debit cards only from VISA and Mastercard.

I think that is a mistake.

**Comment by ianbjacobs** [comment]
If @adrianhopebailie is correct, then I agree there is an issue because that is behavior we sought to enable.

**Comment by marcoscaceres** [comment]
Please see also: https://github.com/w3c/payment-method-basic-card/issues/36#issuecomment-317941633

**Comment by marcoscaceres** [comment]
Questions is, how often does @adrianhopebailie's straw-person occur? and, given that .data is acting as a hint, should the merchant in that case just accept all things in a network (both credit and debit). If it's rare (e.g. 1% of the time), I'm still inclined to "use first matching". 

I've never encountered this in the wild - so I'm curious.

**Comment by ianbjacobs** [comment]
@mattsaxon, @mountainhippo, @lyverovski, could you provide information about the scenario @adrianhopebailie describes, something like "I accept credit cards from all networks but debit cards only from VISA and Mastercard."

Thanks!

Ian

**Comment by marcoscaceres** [comment]
@zkoch, opinions? What are you seeing on the Chrome side or hearing from merchants? Is the situation in https://github.com/w3c/payment-request/pull/581#issuecomment-321785174 rare?

File: index.html [hunk 1]
```diff
@@ -519,8 +519,35 @@
               then <a>throw</a> a <a>TypeError</a>, optionally informing the
               developer that at least one <a>payment method</a> is required.
               </li>
+              <li>Let <var>seenPMIs</var> be an emtpy list.
```
Comment (domenic) [inline index.html]: Nit: "empty" misspelled

File: index.html [hunk 2]
```diff
@@ -519,8 +519,35 @@
...
+                  "payment-method-id#dfn-validate-a-payment-method-identifier">
+                    validate a payment method identifier</a> with
+                    <var>paymentMethod</var>.<a data-lt=
+                    "PaymentMethodData.supportedMethods">supportedMethods</a>.
+                    If it returns false, then throw a <a>RangeError</a>
+                    exception and terminate this algorithm. Optionally, inform
+                    the developer that the payment method identifier is
+                    invalid.
+                  </li>
+                  <li>If <a>seenPMIs</a> contains
```
Comment (domenic) [inline index.html]: `<a>` -> `<var>`

File: index.html [hunk 3]
```diff
@@ -2520,6 +2559,20 @@
                             exception, then <a>abort the update</a> with that
                             exception.
                             </li>
+                            <li>If <a>seenPMIs</a> contains
```
Comment (domenic) [inline index.html]: `<a>` -> `<var>`

**Review by domenic — APPROVED** [review]
EDITORIALLY, lgtm with nits. It seems like people are still discussing the behavior from a normative perspective.

**Comment by marcoscaceres** [comment]
@dontcallmedom, seems the IPR bot is stuck again🕵️‍♀️

**Comment by dontcallmedom** [comment]
Marked as  substantive for IPR from ash-nazg.

**Comment by dontcallmedom** [comment]
@marcoscaceres fixed ; if you have other PRs stuck that way, let me know - this was due to the previous temporary breakage.

**Comment by marcoscaceres** [comment]
Discussed on the editor's call today, we will modify the algorithm to take multiple "supportedNetworks". In the Basic Card spec we will then only use the first matching `supportedTypes`, followed by `supportedNetworks`, to avoid ambiguity.

**Comment by adrianhopebailie** [comment]
> Discussed  on the editor's call today, we will modify the algorithm to take  multiple "supportedNetworks". In the Basic Card spec we will then only  use the first matching supportedTypes, followed by supportedNetworks, to avoid ambiguity.

This seems relevant to Basic Card not Payment Request. Does that mean changes are still needed for PR?

**Comment by marcoscaceres** [comment]
> Does that mean changes are still needed for PR?

Yes, but just the validation of the PMI.

**Comment by marcoscaceres** [comment]
Added missing tests. https://github.com/w3c/web-platform-tests/pull/7021

**Comment by marcoscaceres** [comment]
Blocked awaiting test review.

**Pull request merged**